src/impl.jl

```diff
@@ -0,0 +1,337 @@
+        push!(member_names, b)
+        push!(member_decls, b)
+    end
+    function add_field(__source__, b::Expr)
+        if b.head === :block
+            add_fields(__source__, b)
+        elseif b.head === :const
+            add_field(__source__, b.args[1])
+        elseif b.head === :(::) && b.args[1] isa Symbol
+            push!(member_names, b.args[1])
+            push!(member_decls, b)
+        elseif b.head === :macrocall
+            add_field(__source__, b.args[3])
+        elseif b.head === :function || b.head === :(=) && (b.args[1] isa Expr && b.args[1].head in (:call, :where))
+            # :function, :equals:call, :equals:where are defining functions - inner constructors
+            # we don't want to permit that if it would interfere with us producing them.
+            prevent_inner_constructors &&
+                error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached should not be used on a struct that declares an inner constructor")
+        end
+    end
+    function add_fields(__source__, b::Expr)
+        @assert b.head === :block
+        for field in b.args
+            if field isa LineNumberNode
+                __source__ = field
+            else
+                add_field(__source__, field)
+            end
+        end
+    end
+
+    @assert (struct_decl.args[3].head === :block)
+    add_fields(__source__, struct_decl.args[3])
+    return (member_names, member_decls)
+end
+
+function check_valid_alt_hash_name(__source__, alt_hash_name)
+    isnothing(alt_hash_name) || alt_hash_name isa Symbol || is_expr(alt_hash_name, :.) ||
+        error("$(__source__.file):$(__source__.line): invalid alternate hash function name: `$alt_hash_name`")
+end
+
+isexpr(e, head) = e isa Expr && e.head == head
+isexpr(e, head, n) = isexpr(e, head) && length(e.args) == n
+
+function auto_hash_equals_impl(__source__::LineNumberNode, typ; kwargs...)
+    # These are the default values of the keyword arguments
+    cache=false
+    hashfn=nothing
+    fields=nothing
+
+    # Process the keyword arguments
+    for kw in kwargs
+        @assert kw isa Pair
+        @assert kw.first isa Symbol
+        if kw.first === :cache
+            if !(kw.second isa Bool)
+                error("$(__source__.file):$(__source__.line): `cache` argument must be a Bool, but got `$(kw.second)`")
+            end
+            cache = kw.second
+        elseif kw.first === :hashfn
+            if !(kw.second isa Union{Symbol, Expr, Function})
+                error("$(__source__.file):$(__source__.line): `hashfn` argument must name a function, but got `$(kw.second)`")
+            end
+            hashfn = kw.second
+            check_valid_alt_hash_name(__source__, hashfn)
+        elseif kw.first === :fields
+            if !(kw.second isa Tuple) || !all(f isa Symbol for f in kw.second)
+                error("$(__source__.file):$(__source__.line): invalid `fields` argument: `$(kw.second)`.  Must be of the form `(x, y, z)`.")
```

Comment:
the macro implementation made it look like a single symbol would also be okay - is that not the case? or was the case covering a variable used to pass the names?

Reply:
Yes, a single name will work because of compensating code in the caller.